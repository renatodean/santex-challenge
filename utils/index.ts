import { Item } from "../types/product";

/**
 * From a given product it will return the price as a formated string
 * @param product
 * @param quantity
 * @returns string
 */
export const getPrice = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return ((productPrice * (quantity || 1)) / 100).toFixed(2);
};

/**
 * From a given product it will return the price as number
 * @param product
 * @param quantity
 * @returns number
 */
export const getPriceNumber = (product: Item, quantity?: number) => {
  const { variants } = product || {};
  const productPrice = variants && variants.length > 0 ? variants[0].price : 0;

  return (productPrice * (quantity || 1)) / 100;
};

/**
 * From a given query from router it will return the query value
 * @param query
 * @returns string
 */
export const getQueryValue = (query: any) => {
  if (Array.isArray(query)) {
    return query[0];
  } else {
    return query || "";
  }
};
