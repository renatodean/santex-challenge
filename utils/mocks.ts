export const mockedProducts = {
  products: {
    items: [
      {
        __typename: "Product",
        id: "1",
        name: "Laptop",
        slug: "laptop",
        description:
          "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
        featuredAsset: {
          __typename: "Asset",
          id: "1",
          name: "derick-david-409858-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 129900 },
          { __typename: "ProductVariant", price: 139900 },
          { __typename: "ProductVariant", price: 219900 },
          { __typename: "ProductVariant", price: 229900 },
        ],
      },
      {
        __typename: "Product",
        id: "2",
        name: "Tablet",
        slug: "tablet",
        description:
          'If the computer were invented today, what would it look like? It would be powerful enough for any task. So mobile you could take it everywhere. And so intuitive you could use it any way you wanted — with touch, a keyboard, or even a pencil. In other words, it wouldn’t really be a "computer." It would be Tablet.',
        featuredAsset: {
          __typename: "Asset",
          id: "2",
          name: "kelly-sikkema-685291-unsplash.jpg",
          type: "IMAGE",
          source:
            "https://demo.vendure.io/assets/source/5a/kelly-sikkema-685291-unsplash.jpg",
          preview:
            "https://demo.vendure.io/assets/preview/b8/kelly-sikkema-685291-unsplash__preview.jpg",
        },
        variants: [
          { __typename: "ProductVariant", price: 32900 },
          { __typename: "ProductVariant", price: 44500 },
        ],
      },
    ],
  },
};

export const mockedEmptyProducts = { products: { items: [] } };

export const mockedProduct = {
  product: {
    __typename: "Product",
    id: "1",
    name: "Laptop",
    slug: "laptop",
    description:
      "Now equipped with seventh-generation Intel Core processors, Laptop is snappier than ever. From daily tasks like launching apps and opening files to more advanced computing, you can power through your day thanks to faster SSDs and Turbo Boost processing up to 3.6GHz.",
    featuredAsset: {
      __typename: "Asset",
      id: "1",
      name: "derick-david-409858-unsplash.jpg",
      type: "IMAGE",
      source:
        "https://demo.vendure.io/assets/source/b6/derick-david-409858-unsplash.jpg",
      preview:
        "https://demo.vendure.io/assets/preview/71/derick-david-409858-unsplash__preview.jpg",
    },
    variants: [
      { __typename: "ProductVariant", price: 129900 },
      { __typename: "ProductVariant", price: 139900 },
      { __typename: "ProductVariant", price: 219900 },
      { __typename: "ProductVariant", price: 229900 },
    ],
  },
};

export const mockedStorageState = {
  showSnackbar: true,
  products: [
    {
      product: mockedProduct.product,
      quantity: 3,
    },
  ],
};
