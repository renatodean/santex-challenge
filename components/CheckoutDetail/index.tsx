import React, { useMemo, useState } from "react";
import styled from "@emotion/styled";
import { Item } from "../../types/product";
import { getPriceNumber } from "../../utils";
import CheckoutProductDetail from "../CheckoutProductDetail";

const Container = styled.div`
  background-color: white;
  padding: 16px;
  border-radius: 8px;
  display: grid;
  grid-template-columns: 1fr;
  width: 100%;
  @media (min-width: 900px) {
    grid-template-columns: 2fr 1fr;
  }
`;

const ProductsContainer = styled.div`
  position: relative;
  align-self: center;
  min-width: 256px;
  @media (min-width: 900px) {
    border-right: 0.5px solid rgba(0, 0, 0, 20%);
  }
`;

const PriceInformation = styled.div`
  display: flex;
  flex-direction: column;
  min-width: 200px;
  border-top: 0.5px solid rgba(0, 0, 0, 20%);
  padding-top: 12px;
  align-items: center;
  align-self: center;
  @media (min-width: 900px) {
    border-top: none;
    padding-top: 0;
    padding-left: 16px;
  }
`;

const CheckoutSectionTitle = styled.h3`
  margin-top: 0;
  margin-bottom: 8px;
  text-align: center;
  @media (min-width: 900px) {
    text-align: left;
  }
`;

const PayButton = styled.button`
  background-color: black;
  color: white;
  border: none;
  padding: 10px;
  border-radius: 8px;
  font-size: 20px;
  line-height: 24px;
  margin-top: 10px;
  margin-bottom: 10px;
  cursor: pointer;
  width: 100%;
  &:hover {
    background-color: rgba(1, 1, 1, 70%);
  }
`;

const PaymentSuccessful = styled.h3`
  margin-top: 40px;
`;

const ProductsPayed = styled.p`
  overflow-wrap: anywhere;
`;

interface CheckoutDetailProps {
  products: {
    product: Item;
    quantity: number;
  }[];
}

const CheckoutDetail = ({ products }: CheckoutDetailProps) => {
  const [showPayment, setShowPayment] = useState(false);
  const totalPrice = useMemo(
    () =>
      products.reduce(
        (previous, product) =>
          previous + getPriceNumber(product.product, product.quantity),
        0
      ),
    [products]
  );

  return (
    <>
      <Container>
        <ProductsContainer>
          <CheckoutSectionTitle>Products</CheckoutSectionTitle>
          {products.map((product) => (
            <CheckoutProductDetail key={product.product.name} {...product} />
          ))}
        </ProductsContainer>
        <PriceInformation>
          <CheckoutSectionTitle>
            Total $ {totalPrice.toFixed(2)}
          </CheckoutSectionTitle>
          <PayButton onClick={() => setShowPayment(true)}>Pay</PayButton>
        </PriceInformation>
      </Container>
      {showPayment && (
        <>
          <PaymentSuccessful>
            You have successfully pay the products:
          </PaymentSuccessful>
          <ProductsPayed>{JSON.stringify(products)}</ProductsPayed>
        </>
      )}
    </>
  );
};

export default CheckoutDetail;
