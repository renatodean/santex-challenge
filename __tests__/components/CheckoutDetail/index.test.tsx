import { render, screen, fireEvent } from "@testing-library/react";
import CheckoutDetail from "../../../components/CheckoutDetail";
import { mockedStorageState } from "../../../utils/mocks";

describe("CheckoutDetail", () => {
  it("renders the cartpopup with header", () => {
    render(<CheckoutDetail products={mockedStorageState.products} />);

    expect(screen.getByText("Products")).toBeInTheDocument();
  });

  it("should show products information on pay click", () => {
    render(<CheckoutDetail products={mockedStorageState.products} />);

    const paymentButton = screen.getByText("Pay");
    fireEvent(
      paymentButton,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(
      screen.getByText("You have successfully pay the products:")
    ).toBeInTheDocument();
  });
});

describe("CheckoutDetail snapshot", () => {
  it("renders the cartpopup with products", () => {
    const { container } = render(
      <CheckoutDetail products={mockedStorageState.products} />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the cartpopup without products", () => {
    const { container } = render(<CheckoutDetail products={[]} />);
    expect(container).toMatchSnapshot();
  });
});
