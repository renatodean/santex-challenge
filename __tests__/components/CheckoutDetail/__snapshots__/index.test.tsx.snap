// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutDetail snapshot renders the cartpopup with products 1`] = `
<div>
  <div
    class="css-1l2tmoc"
  >
    <div
      class="css-wbbjgv"
    >
      <h3
        class="css-1v2zqs"
      >
        Products
      </h3>
      <div
        class="css-13y32wl"
      >
        <div
          class="css-17ievcd"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="Laptop"
              class="css-4z07o0"
              data-nimg="fill"
              decoding="async"
              sizes="100vw"
              src="/_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=3840&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fdemo.vendure.io%2Fassets%2Fsource%2Fb6%2Fderick-david-409858-unsplash.jpg&w=3840&q=75 3840w"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
            />
          </span>
        </div>
        <div
          class="css-rq35md"
        >
          <div
            class="css-e165ny"
          >
            3x 
            Laptop
             [$1299.00]
          </div>
          <div
            class="css-1ym2u4i"
          >
            $ 
            3897.00
          </div>
        </div>
      </div>
    </div>
    <div
      class="css-ckvj9p"
    >
      <h3
        class="css-1v2zqs"
      >
        Total $ 
        3897.00
      </h3>
      <button
        class="css-z339wo"
      >
        Pay
      </button>
    </div>
  </div>
</div>
`;

exports[`CheckoutDetail snapshot renders the cartpopup without products 1`] = `
<div>
  <div
    class="css-1l2tmoc"
  >
    <div
      class="css-wbbjgv"
    >
      <h3
        class="css-1v2zqs"
      >
        Products
      </h3>
    </div>
    <div
      class="css-ckvj9p"
    >
      <h3
        class="css-1v2zqs"
      >
        Total $ 
        0.00
      </h3>
      <button
        class="css-z339wo"
      >
        Pay
      </button>
    </div>
  </div>
</div>
`;
